File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
#include <vector>
#include <string>

class Solution {
public:
    std::vector<std::string> letterCombinations(std::string digits) {
        std::vector<std::string> result;
        if (digits.empty()) {
            return result;
        }
        generateCombinations(digits, 0, "", result);
        return result;
    }
private:
    void generateCombinations(const std::string &digits, int index, std::string current, std::vector<std::string> &result) {
        if (index == digits.length()) {
            result.push_back(current);
            return;
        }
        std::string letters = getLetters(digits[index]);
        for (char letter : letters) {
            generateCombinations(digits, index + 1, current + letter, result);
        }
    }
    std::string getLetters(char digit) {
        switch (digit) {
            case '2':
                return "abc";
            case '3':
                return "def";
            case '4':
                return "ghi";
            case '5':
                return "jkl";
            case '6':
                return "mno";
            case '7':
                return "pqrs";
            case '8':
                return "tuv";
            case '9':
                return "wxyz";
            default:
                return "";
        }
    }
};

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.rs
use std::collections::HashMap;


impl Solution {
    pub fn letter_combinations(digits: String) -> Vec<String> {
        let mut result = Vec::new();
        if digits.is_empty() {
            return result;
        }
        let mut map = HashMap::new();
        map.insert('2', "abc");
        map.insert('3', "def");
        map.insert('4', "ghi");
        map.insert('5', "jkl");
        map.insert('6', "mno");
        map.insert('7', "pqrs");
        map.insert('8', "tuv");
        map.insert('9', "wxyz");
        Self::generate_combinations(&digits, 0, &mut String::new(), &mut result, &map);
        result
    }

    fn generate_combinations(
        digits: &str,
        index: usize,
        current: &mut String,
        result: &mut Vec<String>,
        map: &HashMap<char, &str>,
    ) {
        if index == digits.len() {
            result.push(current.clone());
            return;
        }
        if let Some(letters) = map.get(&digits.chars().nth(index).unwrap()) {
            for letter in letters.chars() {
                current.push(letter);
                Self::generate_combinations(digits, index + 1, current, result, map);
                current.pop();
            }
        }
    }
}

